Refactor data deletion

MiguelPires opened the pull request:
This PR refactors the data deletion to the DeletionService and related classes. The second commit is related to #2420 and adds deletion tests for followers and leader with and without exporters. I'm not sure if the issue #2420 is closed by the second commit but I wasn't sure what else to test at the integration level. 

closes #2452
closes #2420

Zelldon commented on broker-core/src/test/java/io/zeebe/broker/logstreams/delete/LeaderLogStreamDeletionTest.java:
```diff
@@ -0,0 +1,145 @@
+    deletionService.delete(POSITION_TO_DELETE);
+
+    // then
+    verify(mockLogBlockIndex, times(1))
```
This is more implementation detail how the deletion works internally in the logstream. I think it is enough to verify that the logstream delete is called with this position.

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/delete/LogStreamDeletionService.java:
```diff
@@ -0,0 +1,57 @@
+
+  public void delete(long position) {
+    position = Math.min(position, getMinimumExportedPosition());
+
```
All lines below should be moved backed to the log stream. The log stream is responsible for the log block index and log storage, we should not modify it outside. It is more an implementation detail of how the log stream handles the deletion.

Then we probably also have no need for this inheritance, if only consist of `
    position = Math.min(position, getMinimumExportedPosition());`

MiguelPires replied:
I think the issue is more on the dependencies. If the logstream does the deletion, then it depends on the exporter service which then depends on the partition service. It would work if we leave the exporter supplier but that was one of the main hacks we wanted to remove with this issue

Zelldon replied:
Hm ? The logstream only gets a upper bound position to which it should delete. No need to know where this came from

MiguelPires replied:
Ah you mean moving only the index and storage delete, not the whole thing, my bad. Yes, that makes sense. About the inheritance, it's only relevant because the way we get the minimum exported position changes from follower to leader, I think that doesn't change right?

Zelldon replied:
Ok :+1: it does not change, but I think we do not need to have inheritance here just use the logstream in both classes and implement the interface. You have the leader deletion service, which implements the interface and resolves the position and calls log stream delete. And we have the FollowerDeletionService which implements the interface, resolves the position and calls the log stream delete.

Zelldon commented on broker-core/src/main/java/io/zeebe/broker/logstreams/delete/FollowerLogStreamDeletionService.java:
```diff
@@ -0,0 +1,87 @@
+  @Override
+  protected long getMinimumExportedPosition() {
+    try {
+      if (snapshotController.getValidSnapshotsCount() > 0) {
```
I think we should check here whether the max count is reached or not right?

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/state/ReplicationController.java:
```diff
@@ -198,9 +199,7 @@ private void validateWhenReceivedAllChunks(
-      if (deleteDataCallback != null) {
-        deleteDataCallback.accept(deletablePositionSupplier.get());
-      }
+      deletionService.delete(deletablePositionSupplier.get());
```
Is it correct that we delete here even if we have not enough valid snapshots?

Zelldon commented on qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/DeleteNoExporterTest.java:
```diff
@@ -0,0 +1,57 @@
+
+  @Test
+  public void shouldDeleteDataOnFollowers() {
+    shouldDeleteDataOnFollowers(clusteringRule, clientRule);
```
Please dont use inheritance in tests :see_no_evil: I have no clue what happens here

Zelldon commented on broker-core/src/main/java/io/zeebe/broker/engine/EngineService.java:
```diff
@@ -137,12 +138,15 @@ private void createAsyncSnapshotDirectorService(final Partition partition) {
         .dependency(
             streamProcessorControllerServiceName,
             snapshotDirectorService.getStreamProcessorInjector())
+        .dependency(deletionServiceName, snapshotDirectorService.getDeletionServiceInjector())
```
Self-answered by continue reading, the director triggers the deletion

MiguelPires commented:
Thanks for the feedback @Zelldon. I refactored the delete code, the tests and added a check in the ReplicationController to only delete on max snapshots

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java:
```diff
@@ -60,7 +60,8 @@ public StateSnapshotController(
               }
             },
-            () -> getPositionToDelete(maxSnapshots));
+            () -> getPositionToDelete(maxSnapshots),
```
Might make more sense to have a callback which is triggered on receiving a valid snapshot. In this callback we then use our information to delete stuff, instead of have multiple suppliers and the DeletionService in the ReplicationController.

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/service/LogStreamService.java:
```diff
@@ -237,26 +235,18 @@ public long getCommitPosition() {
       logStorage.delete(blockAddress);
     } else {
-      LOG.debug(
+      Loggers.LOGSTREAMS_LOGGER.debug(
```
```suggestion
      LOG.debug(
```

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/service/LogStreamService.java:
```diff
@@ -237,26 +235,18 @@ public long getCommitPosition() {
     if (blockAddress != LogBlockIndex.VALUE_NOT_FOUND) {
-      LOG.info(
-          "Delete data from logstream until position '{}' (address: '{}').",
+      Loggers.LOGSTREAMS_LOGGER.info(
```
```suggestion
      LOG.info(
```

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/spi/SnapshotController.java:
```diff
@@ -51,7 +52,7 @@
 
   /** Registers to consumes replicated snapshots. */
-  void consumeReplicatedSnapshots(Consumer<Long> dataDeleteCallback);
+  void consumeReplicatedSnapshots(DeletionService deletionService);
```
This looks weird to me.

Zelldon replied:
But I have no different solution on the fly. I'm not sure whether my problem is the naming or the coupling :sweat_smile: Maybe it would make it a bit better if we rename it to something like `triggerDeletionOnReceivingMaxSnapshotCount` or you split it into two methods to set the deletion service and the other one starts the consuming.

Maybe we can remove the coupling of the replication and the deletion, such that the SnapshotController manages the deletion and is triggered when a valid snapshot was received.

MiguelPires commented:
Hey @Zelldon, I made a new iteration. I refactored the clustered tests to a parameterized test, to avoid duplication, and removed the coupling between the replication and the deletion like you said. I added a NoopSnapshotController because the SnapshotRequester used the ReplicationController without needing an implemented SnapshotController.

Zelldon commented:
I did a bit of clean up, if you fine with them squash and merge please

MiguelPires commented:
bors r+
